File: india_compliance/gst_india/report/hsn_wise_summary_of_purchases/hsn_wise_summary_of_purchases.py
# Copyright (c) 2025, Resilient Tech and contributors
# For license information, please see license.txt
import json

import frappe
from frappe import _
from frappe.utils import getdate

from india_compliance.gst_india.report.hsn_wise_summary_of_outward_supplies.hsn_wise_summary_of_outward_supplies import (
    get_columns,
    process_hsn_data,
    validate_filters,
)
from india_compliance.gst_india.utils.gstr3b.gstr3b_data import GSTR3BInvoices


def execute(filters=None):
    if not filters:
        filters = {}

    validate_filters(filters)

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data


def get_data(filters):
    _class = GSTR3BInvoices(filters)
    invoices = []

    for doctype in ("Purchase Invoice", "Bill of Entry"):
        invoices.extend(_class.get_data(doctype))

    return process_hsn_data(invoices)


@frappe.whitelist()
def get_json(filters, report_name, data):
    from india_compliance.gst_india.report.gstr_1.gstr_1 import get_company_gstin_number

    filters = json.loads(filters)
    report_data = json.loads(data)
    gstin = filters.get("company_gstin") or get_company_gstin_number(filters["company"])

    if not filters.get("from_date") or not filters.get("to_date"):
        frappe.throw(_("Please enter From Date and To Date to generate JSON"))

    fp = "%02d%s" % (
        getdate(filters["to_date"]).month,
        getdate(filters["to_date"]).year,
    )

    gst_json = {"gstin": gstin, "fp": fp}
    gst_json["table18"] = get_hsn_wise_json_data(report_data)

    return {"report_name": report_name, "data": gst_json}


def get_hsn_wise_json_data(report_data):
    data = []

    for hsn in report_data:
        if hsn.get("hsn_code") == "Total":
            continue

        row = {
            "hsn_sc": hsn.get("hsn_code"),
            "uqc": hsn.get("uom"),
            "qty": hsn.get("quantity"),
            "rt": hsn.get("tax_rate"),
            "txval": hsn.get("total_taxable_value"),
            "iamt": 0.0,
            "camt": 0.0,
            "samt": 0.0,
            "csamt": 0.0,
            "isconcesstional": "N",  # TODO: How to identify concession
        }

        if hsn_description := hsn.get("description"):
            row["desc"] = hsn_description[:1000]  # Character limit for GSTR-9 upload

        row["iamt"] += hsn.get("total_igst_amount")
        row["camt"] += hsn.get("total_cgst_amount")
        row["samt"] += hsn.get("total_sgst_amount")
        row["csamt"] += hsn.get("total_cess_amount")

        data.append(row)

    return {"items": data}

File: india_compliance/gst_india/report/hsn_wise_summary_of_purchases/hsn_wise_summary_of_purchases.js
// Copyright (c) 2025, Resilient Tech and contributors
// For license information, please see license.txt

frappe.query_reports["HSN-wise-summary of purchases"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            reqd: 1,
            default: frappe.defaults.get_user_default("Company"),
            on_change: report => {
                report.set_filter_value({
                    company_gstin: "",
                });
                report.refresh();
            },
            get_query: function () {
                return {
                    filters: {
                        country: "India",
                    },
                };
            },
        },
        {
            fieldname: "gst_hsn_code",
            label: __("HSN/SAC"),
            fieldtype: "Link",
            options: "GST HSN Code",
            width: "80",
        },
        {
            fieldname: "company_gstin",
            label: __("Company GSTIN"),
            fieldtype: "Autocomplete",
            get_query() {
                const company = frappe.query_report.get_filter_value("company");
                return india_compliance.get_gstin_query(company);
            },
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            width: "80",
            default: india_compliance.last_month_start(),
            reqd: 1,
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            width: "80",
            default: india_compliance.last_month_end(),
            reqd: 1,
        },
    ],
    onload: report => {
        report.page.add_inner_button(__("Download JSON"), function () {
            var filters = report.get_values();

            frappe.call({
                method: "india_compliance.gst_india.report.hsn_wise_summary_of_purchases.hsn_wise_summary_of_purchases.get_json",
                args: {
                    data: report.data,
                    report_name: report.report_name,
                    filters: filters,
                },
                callback: function (r) {
                    if (r.message) {
                        const args = {
                            cmd: "india_compliance.gst_india.report.hsn_wise_summary_of_outward_supplies.hsn_wise_summary_of_outward_supplies.download_json_file",
                            data: r.message.data,
                            report_name: r.message.report_name,
                        };
                        open_url_post(frappe.request.url, args);
                    }
                },
            });
        });
    },
};
